File: .skills/openclaw-skills/skills/socneo/smart-poller/poller.js
#!/usr/bin/env node
/**
 * Smart Poller — Feishu Task Board Polling Script (Node.js)
 * Periodically checks a Feishu task board and executes tasks assigned to the current agent.
 * Author: socneo
 * Date: 2026-03-16
 */

const https = require('https');
const fs = require('fs');

// Load config from config.json (do NOT hardcode credentials here)
let APP_ID, APP_SECRET, TASK_BOARD_DOC_TOKEN, ASSIGNEE;
try {
    const cfg = JSON.parse(fs.readFileSync('config.json', 'utf8'));
    APP_ID = cfg.app_id;
    APP_SECRET = cfg.app_secret;
    TASK_BOARD_DOC_TOKEN = cfg.doc_token;
    ASSIGNEE = cfg.assignee || 'agent';
} catch (e) {
    console.error('Failed to load config.json:', e.message);
    process.exit(1);
}

// Token cache
let tenantAccessToken = null;
let tokenExpireAt = 0;

/**
 * Obtain (or return cached) Feishu tenant_access_token.
 */
function getTenantAccessToken() {
    return new Promise((resolve, reject) => {
        // Return cached token if still valid
        if (tenantAccessToken && Date.now() < tokenExpireAt) {
            console.log(`  ♻️  Using cached token (${Math.round((tokenExpireAt - Date.now()) / 1000)}s remaining)`);
            resolve(tenantAccessToken);
            return;
        }

        const data = JSON.stringify({ app_id: APP_ID, app_secret: APP_SECRET });
        const options = {
            hostname: 'open.feishu.cn',
            port: 443,
            path: '/open-apis/auth/v3/tenant_access_token/internal',
            method: 'POST',
            headers: { 'Content-Type': 'application/json', 'Content-Length': data.length }
        };

        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                const result = JSON.parse(body);
                if (result.code === 0) {
                    tenantAccessToken = result.tenant_access_token;
                    tokenExpireAt = Date.now() + (result.expire - 300) * 1000; // refresh 5 min early
                    console.log('  ✅ New token obtained');
                    resolve(tenantAccessToken);
                } else {
                    reject(new Error(`Failed to get token: ${result.msg}`));
                }
            });
        });

        req.on('error', reject);
        req.write(data);
        req.end();
    });
}

/**
 * Generic Feishu API request.
 */
function feishuApi(path, token, data = null) {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: 'open.feishu.cn',
            port: 443,
            path: path,
            method: data ? 'POST' : 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${token}`,
                'Content-Length': data ? Buffer.byteLength(JSON.stringify(data)) : 0
            }
        };

        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(body));
                } catch (e) {
                    resolve({ code: res.statusCode, msg: body });
                }
            });
        });

        req.on('error', reject);
        if (data) req.write(JSON.stringify(data));
        req.end();
    });
}

/**
 * Fetch and concatenate all text blocks from the Feishu document.
 */
async function readTaskBoard(token) {
    const result = await feishuApi(`/open-apis/docx/v1/documents/${TASK_BOARD_DOC_TOKEN}/blocks?page_size=200`, token);
    if (result.code !== 0) {
        throw new Error(`Failed to read document: ${result.msg}`);
    }

    let content = '';
    const items = result.data?.items || [];
    for (const block of items) {
        const textFields = ['text', 'heading1', 'heading2', 'heading3', 'bullet'];
        for (const field of textFields) {
            const elements = block[field]?.elements || [];
            for (const elem of elements) {
                const text = elem.text_run?.content || '';
                if (text.trim()) content += text + '\n';
            }
        }
    }
    return content;
}

/**
 * Parse the task board and return pending tasks assigned to ASSIGNEE.
 */
function parseTasks(content) {
    const tasks = [];
    const lines = content.split('\n');
    let currentTask = null;
    const completedTasks = new Set();

    // Pass 1: collect completed task IDs
    for (const line of lines) {
        const completeMatch = line.match(/\[.+\]\s*(TASK-\w+-\d+)/);
        if (completeMatch?.[1]) {
            completedTasks.add(completeMatch[1]);
        }
    }

    // Pass 2: parse pending tasks
    for (const line of lines) {
        const taskMatch = line.match(/\[TASK-(\w+-\d+)\]\s*(.*)/);
        if (taskMatch) {
            if (currentTask &&
                currentTask.assignee === ASSIGNEE &&
                currentTask.status === 'pending' &&
                !completedTasks.has(currentTask.id)) {
                tasks.push(currentTask);
            }
            currentTask = {
                id: `TASK-${taskMatch[1]}`,
                title: taskMatch[2],
                description: '',
                assignee: null,
                status: 'pending',
                priority: 'medium'
            };
            continue;
        }

        // Match assignment line: "Assign: from → to"
        const assignMatch = line.match(/Assign:\s*(\w+)\s*→\s*(\w+)/);
        if (assignMatch && currentTask) {
            currentTask.assignee = assignMatch[2].toLowerCase();
            continue;
        }

        // Match status: "Status: pending"
        const statusMatch = line.match(/Status:\s*(\w+)/);
        if (statusMatch && currentTask) {
            currentTask.status = statusMatch[1].toLowerCase();
            continue;
        }

        // Match priority: "Priority: high"
        const priorityMatch = line.match(/Priority:\s*(\w+)/);
        if (priorityMatch && currentTask) {
            currentTask.priority = priorityMatch[1].toLowerCase();
            continue;
        }

        // Accumulate description
        if (currentTask && line.trim()) {
            currentTask.description += line + '\n';
        }
    }

    // Handle last task
    if (currentTask &&
        currentTask.assignee === ASSIGNEE &&
        currentTask.status === 'pending' &&
        !completedTasks.has(currentTask.id)) {
        tasks.push(currentTask);
    }

    return tasks;
}

/**
 * Append a completion feedback line to the Feishu document.
 */
async function appendFeedback(token, taskId, result) {
    const timestamp = new Date().toLocaleString('en-US', { timeZone: 'Asia/Shanghai' });
    const content = `[${ASSIGNEE} completed] ${taskId} | Time: ${timestamp} | Result: ${result}`;

    const data = {
        children: [{
            block_type: 2, // paragraph
            text: {
                elements: [{ text_run: { content, text_element_style: {} } }],
                style: {}
            }
        }],
        index: -1
    };

    const apiResult = await feishuApi(
        `/open-apis/docx/v1/documents/${TASK_BOARD_DOC_TOKEN}/blocks/${TASK_BOARD_DOC_TOKEN}/children`,
        token,
        data
    );

    return apiResult.code === 0;
}

/**
 * Determine how to handle a task based on its description.
 */
async function executeTask(task) {
    console.log(`  🚀 Executing task: ${task.id} - ${task.title}`);
    const desc = task.description.toLowerCase();

    if (desc.includes('test') || desc.includes('verify')) {
        return 'Test verification successful — polling mechanism is running normally';
    }
    if (desc.includes('search') || desc.includes('find')) {
        return 'Search task received, processing...';
    }
    if (desc.includes('monitor')) {
        return 'Monitoring task started';
    }
    return 'Task received and execution started';
}

/**
 * Main polling function — one full round.
 */
async function pollTasks() {
    console.log('\n' + '='.repeat(60));
    console.log(`📋 Smart Poller [${ASSIGNEE}] — ${new Date().toLocaleString('en-US')}`);
    console.log('='.repeat(60));

    try {
        // Step 1: get access token
        console.log('\n1️⃣  Getting Feishu token...');
        const token = await getTenantAccessToken();

        // Step 2: read task board
        console.log('\n2️⃣  Reading task board...');
        const content = await readTaskBoard(token);
        console.log(`  📄 Document length: ${content.length} chars`);

        // Step 3: parse tasks
        console.log('\n3️⃣  Parsing tasks...');
        const tasks = parseTasks(content);
        console.log(`  📌 Found ${tasks.length} pending task(s) assigned to ${ASSIGNEE}`);

        if (tasks.length === 0) {
            console.log('  ✅ No pending tasks (silent mode: no notification sent)');
            console.log('='.repeat(60) + '\n');
            return;
        }

        // Step 4: execute tasks
        console.log('\n4️⃣  Executing tasks...');
        for (const task of tasks) {
            const result = await executeTask(task);
            console.log(`  ✅ ${task.id}: ${result}`);

            // Step 5: write feedback
            console.log('\n5️⃣  Writing feedback to document...');
            const success = await appendFeedback(token, task.id, result);
            console.log(success ? '  ✅ Feedback written' : '  ❌ Failed to write feedback');
        }

        console.log('\n' + '='.repeat(60));
        console.log('✅ Polling round complete');
        console.log('='.repeat(60) + '\n');

    } catch (error) {
        console.error('❌ Polling failed:', error.message);
    }
}

// Startup
const mode = process.argv[2];

if (mode === '--once') {
    console.log('🔹 Single-run mode');
    pollTasks();
} else {
    const INTERVAL_MS = 5 * 60 * 1000; // 5 minutes default for JS version
    console.log(`🔹 Continuous polling mode — every 5 minutes`);
    console.log(`📡 Press Ctrl+C to stop`);
    pollTasks();
    setInterval(pollTasks, INTERVAL_MS);
}

File: .skills/openclaw-skills/skills/socneo/smart-poller/poller.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Smart Poller - Feishu Task Board Polling Module
Periodically polls a Feishu task board and auto-executes tasks assigned to the current AI agent.
Author: socneo
Version: v1.0
"""

import json
import re
import urllib.request
import urllib.error
from datetime import datetime
from typing import List, Dict, Optional

# Default configuration (overridden by config.json)
DEFAULT_CONFIG = {
    "app_id": "",
    "app_secret": "",
    "doc_token": "",
    "assignee": "agent",
    "poll_interval_minutes": 15,
    "silent_mode": True
}


class FeishuAPI:
    """Feishu API wrapper — handles token management and HTTP requests."""

    def __init__(self, app_id: str, app_secret: str):
        self.app_id = app_id
        self.app_secret = app_secret
        self.token = None
        self.token_expire_at = 0
        self.host = "https://open.feishu.cn"

    def get_tenant_access_token(self) -> str:
        """Obtain (or return cached) Feishu tenant_access_token."""
        if self.token and datetime.now().timestamp() < self.token_expire_at:
            return self.token

        url = f"{self.host}/open-apis/auth/v3/tenant_access_token/internal"
        data = json.dumps({"app_id": self.app_id, "app_secret": self.app_secret}).encode()
        req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"})

        try:
            resp = urllib.request.urlopen(req, timeout=10)
            result = json.loads(resp.read())
            if result.get("code") == 0:
                self.token = result["tenant_access_token"]
                self.token_expire_at = datetime.now().timestamp() + (result["expire"] - 300)
                return self.token
            else:
                raise Exception(f"Failed to get token: {result.get('msg')}")
        except Exception as e:
            raise Exception(f"Request error: {e}")

    def request(self, path: str, data: Optional[Dict] = None) -> Dict:
        """Generic Feishu API request."""
        token = self.get_tenant_access_token()
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}"
        }
        body = json.dumps(data).encode() if data else None
        method = "POST" if data else "GET"

        url = f"{self.host}{path}"
        req = urllib.request.Request(url, data=body, headers=headers, method=method)

        try:
            resp = urllib.request.urlopen(req, timeout=15)
            return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            return {"code": e.code, "msg": e.read().decode()}


class TaskParser:
    """Parses task board content and extracts pending tasks for a given assignee."""

    @staticmethod
    def parse_tasks(content: str, assignee: str) -> List[Dict]:
        """
        Parse the task board document and return pending tasks for the specified assignee.

        Args:
            content: Raw document text
            assignee: Target agent identifier (e.g. 'agent1')

        Returns:
            List of pending task dicts
        """
        tasks = []
        lines = content.split('\n')
        current_task = None
        completed_tasks = set()

        # Pass 1: collect all completed task IDs
        for line in lines:
            match = re.match(r'\[.+\]\s*(TASK-\w+-\d+)', line)
            if match:
                completed_tasks.add(match.group(1))

        # Pass 2: parse tasks
        for line in lines:
            # Match task header: [TASK-xxx-NNN]
            task_match = re.match(r'\[TASK-(\w+-\d+)\]\s*(.*)', line)
            if task_match:
                if (current_task and
                        current_task.get('assignee') == assignee and
                        current_task.get('status') == 'pending' and
                        current_task['id'] not in completed_tasks):
                    tasks.append(current_task)

                current_task = {
                    'id': f"TASK-{task_match.group(1)}",
                    'title': task_match.group(2),
                    'description': '',
                    'assignee': None,
                    'status': 'pending',
                    'priority': 'medium'
                }
                continue

            # Match assignment line: "Assign: from → to"
            assign_match = re.search(r'Assign:\s*(\w+)\s*→\s*(\w+)', line)
            if assign_match and current_task:
                current_task['assignee'] = assign_match.group(2).lower()
                continue

            # Match status: "Status: pending"
            status_match = re.search(r'Status:\s*(\w+)', line)
            if status_match and current_task:
                current_task['status'] = status_match.group(1).lower()
                continue

            # Accumulate description
            if current_task and line.strip():
                current_task['description'] += line + '\n'

        # Handle last task in document
        if (current_task and
                current_task.get('assignee') == assignee and
                current_task.get('status') == 'pending' and
                current_task['id'] not in completed_tasks):
            tasks.append(current_task)

        return tasks


class SmartPoller:
    """Main poller: loads config, polls the task board, and executes tasks."""

    def __init__(self, config_path: str = "config.json"):
        self.config = self.load_config(config_path)
        self.api = FeishuAPI(self.config["app_id"], self.config["app_secret"])
        self.doc_token = self.config["doc_token"]
        self.assignee = self.config["assignee"]
        self.silent_mode = self.config.get("silent_mode", True)

    def load_config(self, path: str) -> Dict:
        """Load configuration from JSON file."""
        try:
            with open(path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except FileNotFoundError:
            print(f"Config file not found: {path}, using defaults.")
            return DEFAULT_CONFIG

    def read_task_board(self) -> str:
        """Fetch and concatenate all text blocks from the Feishu document."""
        result = self.api.request(f"/open-apis/docx/v1/documents/{self.doc_token}/blocks?page_size=200")
        if result.get("code") != 0:
            raise Exception(f"Failed to read document: {result.get('msg')}")

        content = ""
        items = result.get("data", {}).get("items", [])
        for block in items:
            for field in ["text", "heading1", "heading2", "heading3", "bullet"]:
                elements = block.get(field, {}).get("elements", [])
                for elem in elements:
                    text = elem.get("text_run", {}).get("content", "")
                    if text.strip():
                        content += text + "\n"
        return content

    def append_feedback(self, task_id: str, result: str) -> bool:
        """Append a completion feedback line to the Feishu document."""
        timestamp = datetime.now().strftime("%Y/%m/%d %H:%M:%S")
        content = f"[{self.assignee} completed] {task_id} | Time: {timestamp} | Result: {result}"

        data = {
            "children": [{
                "block_type": 2,
                "text": {
                    "elements": [{"text_run": {"content": content, "text_element_style": {}}}],
                    "style": {}
                }
            }],
            "index": -1
        }

        api_result = self.api.request(
            f"/open-apis/docx/v1/documents/{self.doc_token}/blocks/{self.doc_token}/children",
            data
        )
        return api_result.get("code") == 0

    def execute_task(self, task: Dict) -> str:
        """Execute a task based on its description keywords."""
        desc = task.get("description", "").lower()

        if "test" in desc or "verify" in desc:
            return "Test verification successful — polling mechanism is running normally"
        elif "search" in desc or "find" in desc:
            return "Search task received, processing..."
        elif "monitor" in desc:
            return "Monitoring task started"
        else:
            return "Task received and execution started"

    def poll(self) -> Dict:
        """Run a single polling round. Returns a result summary dict."""
        result = {
            "timestamp": datetime.now().isoformat(),
            "tasks_found": 0,
            "tasks_executed": 0,
            "success": True,
            "error": None
        }

        try:
            # Read task board
            content = self.read_task_board()

            # Parse tasks
            tasks = TaskParser.parse_tasks(content, self.assignee)
            result["tasks_found"] = len(tasks)

            if tasks:
                for task in tasks:
                    task_result = self.execute_task(task)
                    self.append_feedback(task["id"], task_result)
                    result["tasks_executed"] += 1
            elif self.silent_mode:
                # Silent mode: no notification when no tasks
                pass

        except Exception as e:
            result["success"] = False
            result["error"] = str(e)

        return result


def main():
    """CLI entry point."""
    import sys

    config_path = sys.argv[1] if len(sys.argv) > 1 else "config.json"
    poller = SmartPoller(config_path)

    # Single-run mode
    if len(sys.argv) > 2 and sys.argv[2] == "--once":
        result = poller.poll()
        print(json.dumps(result, indent=2, ensure_ascii=False))
        return

    # Continuous polling mode
    import time
    interval = poller.config.get("poll_interval_minutes", 15) * 60
    print(f"Smart Poller started — polling every {poller.config.get('poll_interval_minutes', 15)} minutes")
    while True:
        result = poller.poll()
        if not result["success"]:
            print(f"Polling error: {result['error']}")
        time.sleep(interval)


if __name__ == "__main__":
    main()
